### ISSUE TITLE
Recency Sampler state need to be carried from train to val to test

### ISSUE THREAD

**Issue by shenyangHuang** [tgm-team/tgm#157]
In normal TGL workflow, the recency sampler should store information from train set and then carry it to validation set and then carry it to test set.

**Comment by Jacob-Chmura** [comment]
Proposing the following `HookManager` API upgrades, to be discussed.

### HookManager API
```python
register(key: str, hook: DGHook) -> None
register_shared(hook: DGHook) -> None
get_hooks(key: str) -> List[DGHook]
reset_state(key: Optional[str] = None) -> None

scope(key: str, reset_on_exit: bool)
    __enter__() -> List[DGHook]
    __exit__() -> bool
```

### Usage
```python
hook_manager = HookManager() # problem, currently taking dg to extract implicit device
hook_manager.register_shared(nbr_sampler)
hook_manager.register(‘train’, NegatveEdgeSampler(...))
hook_manager.register(‘val’, TGBNegativeEdgeSamplerHook(...))
hook_manager.register(test, TGBNegativeEdgeSamplerHook(...))

for epoch in range(args.epochs)
    with hook_manager.scope('train', reset_on_exit=False) as train_hooks:
        train_loader = DGDataLoader(train_dg, args.bsize, hooks=train_hooks)
        loss = train(train_loader, ...)
    
    with hook_manager.scope('val') as val_hooks:
        val_loader = DGDataLoader(val_dg, args.bsize, hooks=val_hooks)
        val_results = eval(val_loader, ...)
```

### Painpoints
